### ISSUE TITLE
[Build/CI]: Migrate to CI 2.0 workflows 

### ISSUE THREAD

**Issue by torkelo** [grafana/grafana#9633]
Initial poc of ci 2.0 workflows seems very promising

https://github.com/grafana/grafana/tree/circle-ci-workflows

Just need to add a deploy job to make it mergable to master. 

After that we can start looking at reworking how packages & go binaries built (using xgo)

**Comment by bergquist** [comment]
It's time for a new iteration of our build tools and CI/CD pipeline. 

Problems we want to solve
1. CI builds to slow. We build every push to grafana/grafana. Each build takes about 30min. That feedback loop is too long for us and our contributors (we build all pull requests as well). 
2. CD builds are slow. Since we package every successful build, this adds time to every build based of the master branch. 
3. The build tooling used for building and testing Grafana should be easy to use on your local machine. 
4. Our build does not support ARM builds. 
5. We build Linux packages/ci on circle.ci and Windows in Appveyor. (the Appveyor build often fails and is too unstable for CI)
6. Unit tests only run on sqlite. We have no automated testing of mysql/(mariadb?)/postgres
7. We have no automated integration tests at all.
7. We only support master and release builds. Not beta/nighly.
8. The release process requires many manual steps. We should reduce this to max 5. (more info on our internal wiki)

Suggested solutions

**Comment by torkelo** [comment]
To tackle the problem of speed I started work on using ci workflows and also using CI standard containers for node & go to run the build & unit tests. This increase build time by multiple factors.

See changes to circle.yaml and 
https://github.com/grafana/grafana/tree/circle-ci-workflows
Updated
https://github.com/grafana/grafana/blob/circle-ci-workflows/circle.yml
And the frontend/backend test sh scripts 

These steps could then store assets (frontend/backend binary) to be used by the latest steps. 

## Cross compiling
I have had some success testing https://github.com/karalabe/xgo  for cross compiling linux, windows & darwin (osx) & arm builds. The only problem is that that the docker container used by xgo for the cross compiling is based on debian 16, this will cause linux binaries to depend on a newer version of glibc than exists on older linux systems and you will get a binary that does not work on debian 14 or RHEL 6 / Centos 6. (you will get an error like version `GLIBC_2.14' not found when starting grafana-server). This could be solved by forking https://github.com/karalabe/xgo and modifying it's base docker container to use debian 10.04 as base or debian 6 Squeeze. 

The alternative would be to build linux binary in debian 16 and see how many still run CentOS 6 / RHEL 6 .

**Comment by fg2it** [comment]
@torkelo @bergquist @xlson `xgo` doesn't seems necessary. `grafana-server` and `grafana-cli` seems to build fine (at least for armv6/v7/64) with the CC env variable for `go build`. If fact, this is precisly how the option [`-cc=gcc-xxx`](https://github.com/grafana/grafana/blob/338655dd37fd3105eb6b7e1831a2d53be0e638b0/build.go#L56) of `build.go` do it : [creates env variables](https://github.com/grafana/grafana/blob/338655dd37fd3105eb6b7e1831a2d53be0e638b0/build.go#L441-L455) (`-cc`/CC, `-cgo-enable`/CGO_ENABLE and `-goarch`/GOARCH-GOARM ) to give them to [`go build`](https://github.com/grafana/grafana/blob/338655dd37fd3105eb6b7e1831a2d53be0e638b0/build.go#L407). So you can select the gcc version you want.

`-phjs`, `-pkg-arch` are not needed for `grafana-server` and `grafana-cli` (first one supplies the right binary and not the one from fetched by node, the second is used in `grunt` to have the correct arch name in the `tar.gz`), and `-cxx` is in fact not needed at all, sorry for this one 😄 .

For reference, that was introducted in this [PR](6068)

**Comment by torkelo** [comment]
it is needed for cross platform windows & darwin builds since Grafana uses CGO. Building for other OS platforms with CGO (not just other processors) is not straight forward and requires a very complex environment with headers and object files specifically for that platform, which is what xgo does.

**Comment by fg2it** [comment]
@torkelo you surely need a working toolchain and not only the compiler. I check for windows based on small changes of my last [branch](/fg2it/grafana-on-raspberry/tree/fast-build) (it rebuilds only `grafana-server` and `grafana-cli`, and reuses official linux x64 tgz) and everything works smoothly (cross build and run on windows10) just "apt-get installing" `gcc-mingw-w64` and using adding `GOOS=windows GOARCH=amd64 CGO_CFLAGS=-D_WIN32_WINNT=0x0400` for `go build`. Will try a toolchain for osx, maybe it is less easy.
Anyway, almost as is, this branch can build armv6, v7, aarch64 and win64.

**Comment by fg2it** [comment]
@torkelo modified branch [here](/fg2it/grafana-on-raspberry/tree/fast-build-all) seems to works fine. Build on osx succeeded but is untested yet. Toolchain for osx is arguably less simple than win64 but is still easy (cf. this [repo](/docker/golang-cross) of docker which is quite similar to xgo for the osx toolchain itself, but stripped of all unneeded other tools). Will try to test the osx built (which by the way is only a zip and not yet a homebrew formula).
The whole thing, (armv6/v7/64/win64/asx) took less than 25 min cpu time to build (osx toolchain included) according to travis, and less than 10 min elapse (with 3 concurrent builds).

**Comment by torkelo** [comment]
but why not use xgo? it just a prepped docker container with all of those dependencies. Sure we can redo that in our own build container but what is the advantage?

**Comment by fg2it** [comment]
That is a different concern from the "troublesome" initial one but, fair enough, that certainly is a relevant question : 
- reduced dependencies (you just require what is needed, which is a small subset of what xgo uses)
- direct control on each of those dependencies (if for some reason, a toolchain for a target need to be changed, you can do so easily, on your own)
- higher auditability (if a problem arise in toolchain, it will be much easier to track and fix)
- lighter (contributor don't need to pull a 4.7Go gigantic docker image ; depending on bandwidth/cpu trade off, it could also likely be faster, eg. only pulling xgo took me over 10min vs 7 min to build grafana --and the toolchain-- for osx)
- better parallelism (each target are build in the same container in xgo; granted, building grafana binaries is not the most expansive)
- use golang official binaries (rebuild in xgo)

**Comment by torkelo** [comment]
ok, you make some good arguments. but the issue of the large docker container sounds like a non issue, it will be cached on the build server and will only affect users who want to do custom cross platform builds (very small subset of users/contributors). 

 If you have anything you can contribute that would be great. Otherwise I see using xgo as massive time saver to not have to deal with all that complexity that it solves. Not sure suing a custom container compared to xgo is that much of a "reducing dependencies", seems like like reinventing the wheel.

**Comment by torkelo** [comment]
did not know xgo recompiles the go compiler, wonder why. Also strange why xgo container is so large, maybe a cleanup is missing? We would have to create a build container that is almost identical (in features at least).

**Comment by fg2it** [comment]
xgo contains something like 8 natively gcc supported toolchain, android and ios ndk are supported for 3 differents arch, std golang lib are build for all supported arch, I guess that even with cleaning, you end up with a large base image.
Reducing the dependencies is one `apt-get` and 6 lines in bash. After that, you just use golang cross build feature. I don't see the need for a dedicated container like xgo to simply build grafana binaries. One the toolchain are setup, golang provides everything needed for crossbuild with `GOOS` and the like. The needed toolchain could easly be setup in [grafana-build-container](https://github.com/grafana/grafana-build-container) (well, not as easily on the current centos:6.6 base image that is lacking some toolchains but on eg ubuntu:14.04 or latter).
Beside, It seems that crossbuilding the binaries (beeing with xgo or needed-toolchain+golang) won't be the dev consuming part, I think it will be integration with current build-flow.

**Comment by torkelo** [comment]
> Reducing the dependencies is one apt-get and 6 lines in bash. 

That sounds really great! Can you share that? (and also the build commands). The reason for centos 6.6 is to create linux binaries that are compatible with older linux distros (GLIBC version). Might be able to use ubuntu:14.04 instead not sure. 

If installing the tooling and prepping   ourexisting build container for cross platform build is as easy as you say I agree, xgo is probably overkill.

**Comment by xlson** [comment]
If we can use something like this it would simplify a lot. I ran the build on my dev box in 4 minutes. Will test the binaries when I have access to machines. When it comes to building with an old GLIBC version I guess we could have two separate builds if getting all of them out of the same container ends up being too much work.

**Comment by torkelo** [comment]
@fg2it do you have any ideas for getting a glibc v2.14 compatible linux build using your ubuntu container or do we need to convert it to centos 6.6 for that? Our other (current) plan is to use your build container for cross platform builds but keep the centos 6.6 for the main amd64 build in order to keep the binary compatability with older linux distros. Its not ideal to have to split the CI build into two build containers like this... :(

**Comment by fg2it** [comment]
@torkelo glibc 2.14 is probably a typo since this is what is used by ubuntu:14.04. So I guess you want an older glibc to keep the same dep as the current build.

Here is what I tried. I chose what seems the easiest and the fastest to do and start from ubuntu:14.04 to build a gcc toolchain with an old glibc.
There is a gcc-4.4 (in fact a 4.4.7, the same as centos:6.6) directly available in ubuntu:14.04 but it is stubborn and link with glibc 2.14 (Most likely it was build on the system and use the default glibc 2.14) which ruins the purpose.
So I went for crossbuild-NG, ending with the last commit I pushed on [#11473](https://github.com/grafana/grafana/pull/11473#issuecomment-379926356).

It might be possible to do it the other way starting from centos6.6 and building the missing toolchains (I didn't recheck that but I think it is *all* the other toolchains), but this is probably more work and I am not sure everything will build nicely for arm.

By the way, slightly on an other topic, the frontend is currently build for each platform. It might be worth to build it once, then build the binaries for all the platforms and package several time. Don't know how this would fit with circle ci. That is what I try in this [branch](https://github.com/fg2it/grafana-on-raspberry/tree/fast-build-all) of my grafana-on-raspberry repo (I cheated a bit since it does even build the frontend: it reuses the official one).

That is all for me tonight. Hope it helps.

**Comment by torkelo** [comment]
> By the way, slightly on an other topic, the frontend is currently build for each platform. It might be worth to build it once, then build the binaries for all the platforms and package several time. Don't know how this would fit with circle ci. 

building the frontend only once should definitely be part of working on cross platform CI builds. It is the most time consuming part of the build and only needs to be done once. It is only done once per package (tar/rpm/deb) so no reason to do for each platform.

**Comment by fg2it** [comment]
@torkelo Good. That is what is done in [build.py](https://github.com/grafana/grafana/pull/11473#issuecomment-381437171) in the commit.
The current split of the packaging logic between grunt (tgz/zip) and build.go (rpm/deb) wasn't nice for building all the packages. So to avoid changing too many things I simply add the build.py script (and made a small changes in grunt tasks). 
The python script and the modified grunt tasks are all that is needed to build the current master in the image based on Dockerfile.builder.
Beside integration to your CI, there is the problem of PhantomJS : they don't provide binaries for armv7/arm64. So in my PR I used my build. I would probably be better if you use your own build of PhantomJS for armv7/arm64.

**Comment by torkelo** [comment]
We will move from phantomjs to headless chrome and remove this as included  binary and instead make this into an optional dependency (that deb or rpm may install automatically or user will have to do it manually)

**Comment by torkelo** [comment]
@xlson is working on this (based your work) , modifying our current build container so it's based on yours. 
https://github.com/grafana/grafana-build-container 

Not sure how far he has gotten or if there is more you can help with. He is on vacation this week so we will have to wait until next week.

**Comment by xlson** [comment]
@fg2it Thanks again for putting together the POC. I think we should try to stick to build.go as to not introduce another language in the core stack for Grafana development. Unless there is something I'm missing that would make it more beneficial to use python. Was the choice of python one you made out of familiarity or was it something else that made you go for python?

My plan, for now, is to start by replacing our build container with yours so that we can get it into the existing pipeline and start using it for master builds. After that, I'm going to document our build process and what changes are needed so that build.go can do what's necessary and start modifying the build.go/grunt workflow to make cross-platform builds possible.

**Comment by mjtrangoni** [comment]
Hi, I was digging a little bit and figured this out with [checksec](github.com/slimm609/checksec.sh). See,
``` 
$ checksec --file grafana-server
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	FORTIFY	Fortified Fortifiable  FILE
Partial RELRO(yellow)   Canary found      NX enabled    No PIE(red)          No RPATH   No RUNPATH
```
This can be fixed with `-buildmode=pie`, see `go help buildmode`. 

```
checksec --file ./bin/grafana-server
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	FORTIFY	Fortified Fortifiable  FILE
Partial RELRO   Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH 
```

**Comment by fg2it** [comment]
@xlson I wanted to avoid go, so native support of crossbuild config could be used. Having `build.go`, you cannot crossbuild with the usual `GOARCH=.... go run build.go ...` because `go run` will use the env settings to build `build.go` (so the need for `-cc`, `-goarch` and the like) . So, the first reason is not to use python, but to not use go. That being said, I am indeed more at ease with python. 
I understand perfectly that introducing a new language is a concern. But, I think it would be better to have something else than a "go script" to build.
I also think that build.go could use some refactor (eg `build-srv` and `build-server`), so starting from scratch was an appealing idea.

**Comment by torkelo** [comment]
Maybe bazel could be worth exploring, I know next to nothing about it , and don’t want to introduce an external build tool for normal scenarios. But I know Tom sing super high praise for it and how it simplifies some build stuff and seems to have powerful caching and optimizing feature,
https://bazel.build/

**Comment by SeppPenner** [comment]
Looks good, I'm looking forward to get the https://github.com/grafana/grafana-docker/issues/117 Dockerfile for ARM support soon :)
